File: src/org/jetbrains/haskell/sdk/HaskellSdkConfigurable.java
package org.jetbrains.haskell.sdk;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.projectRoots.AdditionalDataConfigurable;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.projectRoots.SdkAdditionalData;
import com.intellij.openapi.projectRoots.SdkModificator;

import javax.swing.*;

public final class HaskellSdkConfigurable implements AdditionalDataConfigurable {
    private final HaskellSdkConfigurableForm myForm;

    private Sdk mySdk;

    public HaskellSdkConfigurable() {
        myForm = new HaskellSdkConfigurableForm();
    }

    public void setSdk(Sdk sdk) {
        mySdk = sdk;
    }

    public JComponent createComponent() {
        return myForm.getContentPanel();
    }

    public boolean isModified() {
        return myForm.isModified();
    }

    public void apply() {
        HaskellSdkAdditionalData newData = new HaskellSdkAdditionalData(myForm.getGhcOptions(), myForm.getCabalPath());

        final SdkModificator modificator = mySdk.getSdkModificator();
        modificator.setSdkAdditionalData(newData);
        ApplicationManager.getApplication().runWriteAction(new Runnable() {
            public void run() {
                modificator.commitChanges();
            }
        });
        myForm.setModified(false);
    }

    public void reset() {
        SdkAdditionalData data = mySdk.getSdkAdditionalData();
        HaskellSdkAdditionalData ghcData;
        if (data != null) {
            if (!(data instanceof HaskellSdkAdditionalData))
                return;
            ghcData = (HaskellSdkAdditionalData) data;
        } else {
            ghcData = null;
        }
        boolean modified = false;
        String initialGhcOptions = "-W";
        String ghcOptions = ghcData == null ? initialGhcOptions : ghcData.getGhcOptions();
        String cabalPath  = ghcData == null ? "" : ghcData.getCabalPath();
        myForm.init(ghcOptions, cabalPath);
        myForm.setModified(modified);
    }

    public void disposeUIResources() {
    }
}

File: plugin/src/org/jetbrains/haskell/sdk/HaskellSdkConfigurable.kt
package org.jetbrains.haskell.sdk

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.projectRoots.AdditionalDataConfigurable
import com.intellij.openapi.projectRoots.Sdk
import com.intellij.openapi.projectRoots.SdkAdditionalData
import com.intellij.openapi.projectRoots.SdkModificator
import org.jetbrains.haskell.util.OSUtil
import java.io.File
import javax.swing.*

class HaskellSdkConfigurable : AdditionalDataConfigurable {
    private val form: HaskellSdkConfigurableForm = HaskellSdkConfigurableForm()

    private var mySdk: Sdk? = null

    override fun setSdk(sdk: Sdk?) {
        mySdk = sdk
    }

    override fun createComponent(): JComponent {
        return form.getContentPanel()
    }

    override fun isModified(): Boolean {
        return form.isModified
    }

    override fun apply() {
        val newData = HaskellSdkAdditionalData(
                form.getGhciPath(),
                form.getGhcpkgPath(),
                form.getCabalPath())

        val modificator = mySdk!!.sdkModificator
        modificator.sdkAdditionalData = newData
        ApplicationManager.getApplication()!!.runWriteAction(object : Runnable {
            override fun run() {
                modificator.commitChanges()
            }
        })
        form.isModified = false
    }

    override fun reset() {
        val sdk = mySdk!!
        val data = sdk.sdkAdditionalData

        if (data != null) {
            if (data !is HaskellSdkAdditionalData) {
                return
            }
            val ghcData: HaskellSdkAdditionalData = data
            val ghciPath = ghcData.ghciPath ?: ""
            val ghcPkgPath = ghcData.ghcPkgPath ?: ""
            val cabalPath = ghcData.cabalPath ?: ""

            form.init(ghciPath, ghcPkgPath, cabalPath)
        } else {
            val file = File(sdk.homePath)
            val version = extractVersion(file.name)
            val parent = file.parent
            form.init(
                    File(parent, OSUtil.getExe("ghci-" + version)).toString(),
                    File(parent, OSUtil.getExe("ghc-pkg-" + version)).toString(),
                    File(parent, OSUtil.getExe("cabal")).toString())
        }

        form.isModified = false
    }

    private fun extractVersion(name: String) : String {
        val trimmedName = OSUtil.removeExtension(name)

        if (trimmedName == "ghc") {
            return ""
        }
        if (trimmedName.startsWith("ghc-")) {
            return name.substring(4)
        }
        return ""
    }

    override fun disposeUIResources() {
    }

}
